lib/model/recent_senders.dart

```diff
@@ -0,0 +1,174 @@
+import 'package:collection/collection.dart';
+
+import '../api/model/model.dart';
+import 'algorithms.dart';
+
+class MessageIdTracker {
+  final List<int> _ids = [];
+
+  void add(int id) {
+    if (_ids.isEmpty) {
+      _ids.add(id);
+    } else {
+      int i = possibleIndexIn(_ids, id);
+      if (i >= 0) { // the [id] already exists, so do not add it.
+        return;
+      }
+      i = -i - 1; // change the index back to 0-based for readability.
+      if (i == _ids.length) { // the [id] should be added to the end of the list.
+        _ids.add(id);
+      } else {
+        _ids.insert(i, id);
+      }
+    }
+  }
+
+  // TODO: remove
+
+  /// The maximum id in the tracker list.
+  ///
+  /// Returns -1 if the tracker list is empty.
+  int get maxId => _ids.isNotEmpty ? _ids.last : -1;
+
+  /// Getter for the tracked message IDs.
+  ///
+  /// This is intended for testing purposes only.
+  List<int> get idsForTesting => _ids;
+
+  @override
+  bool operator == (covariant MessageIdTracker other) {
+    if (identical(this, other)) return true;
+
+    return _ids.equals(other._ids);
+  }
+
+  @override
+  int get hashCode => Object.hashAll(_ids);
+}
+
+/// A data structure to keep track of stream and topic messages.
+///
+/// The owner should call [clear] in order to free resources.
+class RecentSenders {
+  // streamSenders[streamId][senderId] = IdTracker
+  final Map<int, Map<int, MessageIdTracker>> _streamSenders = {};
+
+  // topicSenders[streamId][topic][senderId] = IdTracker
+  final Map<int, Map<String, Map<int, MessageIdTracker>>> _topicSenders = {};
+
+  /// Getter for the stream senders.
+  ///
+  /// This is intended for testing purposes only.
+  Map<int, Map<int, MessageIdTracker>> get streamSendersForTesting => _streamSenders;
+
+  /// Getter for the topic senders.
+  ///
+  /// This is intended for testing purposes only.
+  Map<int, Map<String, Map<int, MessageIdTracker>>> get topicSendersForTesting => _topicSenders;
+
+  void clear() {
+    _streamSenders.clear();
+    _topicSenders.clear();
+  }
+
+  int _latestMessageIdOfSenderInStream({required int streamId, required int senderId}) {
+    return _streamSenders[streamId]?[senderId]?.maxId ?? -1;
+  }
+
+  int _latestMessageIdOfSenderInTopic({
```

Comment:
Haven't touched this part yet. Should I move `compareByRecency` to its caller and thus change these two to public? Right now they're not used outside of this class.

Reply:
Oh, and missed this question in my sweep just above. Yeah, let's move that function to near where it's called (and/or inline it into its caller), and make these two public.

That'll be part of the second PR that gets stacked atop this one, since this is in the recent-senders code.

Reply:
(Adding another comment now, as part of "a review" this time, just to make this show up in the main PR page's chronological list.)